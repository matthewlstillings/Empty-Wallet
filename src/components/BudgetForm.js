import React from 'react';
import {connect} from 'react-redux';
import {startAddIncome, addIncome, subtractIncome} from '../actions/budget';


export default class BudgetForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            budget: props.income ? (props.income.amount / 100).toString() : '',
            error: 0
        }
    }
    onAdditionChange = (e) => {
        const budget = e.target.value;
        if (budget.match(/^\d{1,}(\.\d{0,2})?$/g) || !budget) {
            this.setState(() => ({budget}));
        }
    };
    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit({
            budget: parseFloat(this.state.budget, 10) * 100 
        });
        this.setState((prevState) => ({error: prevState.error + 1 }));
        console.log(this.state.error);
    };
    resetError = () => {
        this.setState(()=>({error: 0}));
    }
    componentDidMount = () => {
        let intervalId = setInterval(() => {this.resetError()}, 5000);
        try {
            const json = localStorage.getItem('error');
            const error = JSON.parse(json);
            if (error) {
                this.setState(()=>({error}))
                console.log('Found Data');
            }
    
        } catch (e) {
            //Does nothing
        }
    }
    componentDidUpdate = () => { //logs when state updates
            const json = JSON.stringify(this.state.error);
            localStorage.setItem('error', json);
            console.log('saving data');
    };   
    render() {
        return (
            <div className="budget-form__container">
                <form onSubmit={this.onSubmit} className="budget-form">
                    <input 
                        type="text"
                        value={this.state.budget}
                        onChange={this.onAdditionChange}
                        disabled={this.state.error > 2}
                        className="budget-form__description"
                        placeholder="Enter a new budget amount"
                    />
                    <button className="budget-form__btn" disabled={this.state.error > 2}>Update Budget</button>
                </form>
                {this.state.error > 2 && <h3 className="expense-error">You're doing that too much, calm down!</h3>}

            </div>
        );
    }
};