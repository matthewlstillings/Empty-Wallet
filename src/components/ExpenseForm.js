import React from 'react';
import moment from "moment";
import {SingleDatePicker} from 'react-dates';


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            notes: props.expense ? props.expense.notes : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt): moment(),
            focused: false,
            error: ''
        };
    }
    
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=> ({description}));
    };
    onNotesChange = (e) => {
        const notes = e.target.value;
        this.setState(()=> ({notes}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (amount.match(/^\d{1,}(\.\d{0,2})?$/g) || !amount) {
            this.setState(() => ({amount}));
        }
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt}));
        }
        
    };
    onFocusChange = ({focused}) => {
        this.setState(()=>({focused}));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(()=>({error: 'error'}));
        } else {
            this.setState(()=>({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    render () {
        return (
            <div>
                {this.state.error === 'error' && <p>Please provide a description and amount</p>}
                <form onSubmit={this.onSubmit}>
                   <input 
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                   />
                   <input 
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                   />
                   <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.focused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=> false}
                   />
                   <textarea
                    placeholder="Notes"
                    value={this.state.notes}//This might need to be changed to "notes"
                    onChange={this.onNotesChange}
                   >
                   </textarea> 
                   <button>Add Expense</button>
                </form>
            </div>
        )
    }
}