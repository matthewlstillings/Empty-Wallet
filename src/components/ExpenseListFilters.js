import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
    inputFocus: false
  };
  onInputFocus = () => {
    this.setState((prevState) => ({ inputFocus: !prevState.inputFocus }));
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      
        <div className="budget-dashboard__filters">
          <DateRangePicker
            startDate={this.props.filters.startDate}
            endDate={this.props.filters.endDate}
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            showClearDates={true}
            numberOfMonths={1}
            isOutsideRange={() => false}
            showDefaultInputIcon= {true}
          />
          <div className="filters__container">
            <input
              type="text"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              className="budget-dashboard__text-filter"
              placeholder={this.state.inputFocus ? "" : "Search Expenses"}
              onFocus={this.onInputFocus}
              onBlur={this.onInputFocus}
              
            />
            <div className="budget-dashboard__sort__container">
            <p>Sort:</p>
              <select
                value={this.props.filters.sortBy}
                onChange={this.onSortChange}
                className="budget-dashboard__sort"
              >
                <option className="sort-drop" value="date">Date</option>
                <option className="sort-drop" value="amount">Amount</option>
              </select>
            </div>
          </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);