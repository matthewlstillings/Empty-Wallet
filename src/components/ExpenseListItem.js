import React from 'react';
import {Link} from 'react-router-dom';




export const Expense = ({id, dispatch, description, amount, createdAt}) => (
        <li>
            <Link to={`/edit/${id}`}>
                <h3>{description}</h3>
            </Link>
            <p>{amount}</p>
            <p>{createdAt}</p>
        </li>
);



export default Expense; //Maybe Fix this later?