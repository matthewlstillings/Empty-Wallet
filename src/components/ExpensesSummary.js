import React from 'react';
import {connect} from 'react-redux';
import total from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpenseSummary = (props) => {
    const expenseWord = props.expenseCount === 1 ? 'expense' : 'expenses';
    const expenseSummaryFormatted = numeral(props.expensesSummary).format('$0,0.00'); //not dividing here might break other places, trying dividing by 100 here instead of selector
    return (
        <div>
            <h3>You are viewing {props.expenseCount} {expenseWord}, totalling {expenseSummaryFormatted}.</h3>
        </div>
    );
};



const mapStateToProps = (state) => {
    return {
        expensesSummary: total(selectExpenses(state.expenses, state.filters)),
        expenseCount: selectExpenses(state.expenses, state.filters).length
    };
};

export default connect(mapStateToProps)(ExpenseSummary);