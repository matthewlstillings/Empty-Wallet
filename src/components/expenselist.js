import React from 'react';
import {connect} from 'react-redux';
import Expense from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <ul className="expense-list">
        {
          props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ) : (
                props.expenses.map((expense)=> {
                   return <Expense 
                         key={expense.id}
                         {...expense}
                    />;
                })
            )
        }
        </ul>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
        budget: state.budget
    };
};

export default connect(mapStateToProps)(ExpenseList);

