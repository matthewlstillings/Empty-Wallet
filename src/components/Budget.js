import React from 'react';
import {connect} from 'react-redux';
import {startAddIncome} from '../actions/budget';
import ExpenseSummary from './ExpensesSummary';
import BudgetForm from './BudgetForm';


export class BudgetManager extends React.Component {
    onAddSubmit = (income) => {
        this.props.startAddIncome(income);
        window.location.reload();
    };
   
    render() {
        return (
            <div>
                <h1 className="add-expense__title">Budget Manager</h1>
                <BudgetForm
                    onSubmit={this.onAddSubmit}
                />
            </div>
        );
    }
};


const mapDispatchToProps = (dispatch) => ({
    startAddIncome: (income) => dispatch(startAddIncome(income))
});



export default connect(undefined, mapDispatchToProps)(BudgetManager);