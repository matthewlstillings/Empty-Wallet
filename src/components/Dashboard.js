import React from 'react';
import ExpenseList from './expenselist';
import ExpenseSummary from './ExpensesSummary';
import ExpenseListFilters from './ExpenseListFilters';

const BudgetDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseSummary />
        <ExpenseList />
    </div>
);

export default BudgetDashboardPage;
