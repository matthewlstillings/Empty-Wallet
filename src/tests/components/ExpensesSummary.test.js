import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseSummary} from '../../components/ExpensesSummary';

test('should render expense with 1 expense', ()=> {
    const wrapper = shallow(<ExpenseSummary expensesSummary={145} expenseCount={1} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render expenses with multiple items', ()=> {
    const wrapper = shallow(<ExpenseSummary expensesSummary={145241} expenseCount={3} />);
    expect(wrapper).toMatchSnapshot();
});

