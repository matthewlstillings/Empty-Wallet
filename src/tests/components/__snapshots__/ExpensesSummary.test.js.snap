// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense with 1 expense 1`] = `
<div>
  <h3>
    You are viewing 
    1
     
    expense
    , totalling 
    $145.00
    .
  </h3>
</div>
`;

exports[`should render expenses with multiple items 1`] = `
<div>
  <h3>
    You are viewing 
    3
     
    expenses
    , totalling 
    $145,241.00
    .
  </h3>
</div>
`;
